File: src/lowlevel/server/icefsoperators/icefsStatFS.go
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-11 07:18:32
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-18 14:58:59
 * @FilePath: /icefs/src/lowlevel/server/icefsoperators/icefsStatFS.go
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
package icefsoperators

import (
	"context"
	"icefs-server/icefserror"
	pb "icefs-server/icefsgrpc"
	"icefs-server/icefsthrift"
	"syscall"
)

func (s *IcefsServer) doIcefsStatFS(fakeInode uint64, statvfsStructBuilder StatvfsStructBuilder) (status int32, statvfs any) {
	var err error
	var statfs syscall.Statfs_t
	s.inodeCacheLock.RLock()
	inode := s.getIcefsInode(fakeInode)
	if inode == nil {
		s.inodeCacheLock.RUnlock()
		status = icefserror.ICEFS_BUG_ERR
		return
	}
	inode.inodeLock.RLock()
	s.inodeCacheLock.RUnlock()
	err = syscall.Fstatfs(inode.fd, &statfs)
	inode.inodeLock.RUnlock()
	if err != nil {
		status = icefserror.IcefsStdErrno(err)
		return
	}
	status = icefserror.ICEFS_EOK
	statvfs = statvfsStructBuilder(&statfs)
	return
}

func (s *IcefsGRpcServer) DoIcefsStatFS(ctx context.Context, req *pb.IcefsStatFSReq) (*pb.IcefsStatFSRes, error) {
	var res pb.IcefsStatFSRes
	var statvfs any

	res.Status, statvfs = s.server.doIcefsStatFS(req.Inode, GRpcStatvfsStructBuilder)
	if res.Status == icefserror.ICEFS_EOK {
		res.Statvfs = statvfs.(*pb.StatvfsStruct)
	}

	return &res, nil
}

func (s *IcefsThriftServer) DoIcefsStatFS(ctx context.Context, req *icefsthrift.IcefsStatFSReq) (*icefsthrift.IcefsStatFSRes, error) {
	var res icefsthrift.IcefsStatFSRes
	var statvfs any

	res.Status, statvfs = s.server.doIcefsStatFS(uint64(req.Inode), ThriftStatvfsStructBuilder)
	if res.Status == icefserror.ICEFS_EOK {
		res.Statvfs = statvfs.(*icefsthrift.StatvfsStruct)
	}

	return &res, nil
}

File: src/lowlevel/client/icefsoperators/icefsStatFS.cpp
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-30 04:19:29
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-21 02:41:05
 * @FilePath: /icefs/src/lowlevel/client/icefsoperators/icefsStatFS.cpp
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
#include <stdio.h>

#include <string>

#include "icefsClient.hpp"
#include "icefsClientUtils.hpp"

void IcefsClient::DoIcefsStatFS(fuse_req_t fuseReq, fuse_ino_t inode) {
  ICEFS_PR_FUNCTION;

  switch (this->clientConfig.linkType) {
    case ICEFS_LINK_USE_GRPC: {
      icefsgrpc::IcefsStatFSReq req;
      icefsgrpc::IcefsStatFSRes res;
      grpc::ClientContext ctx;

      struct statvfs statData;
      req.set_inode(inode);

      grpc::Status status = gRpcClient->DoIcefsStatFS(&ctx, req, &res);
      if (status.ok() && !res.status()) {
        IcefsGRpcFillStatvfsIn(&statData, res.statvfs());
        fuse_reply_statfs(fuseReq, &statData);
      } else {
        fuse_reply_err(fuseReq, res.status() ? res.status() : EIO);
        ICEFS_PR_ERR_STATUS;
      }
      break;
    }

    case ICEFS_LINK_USE_THRIFT: {
      icefsthrift::IcefsStatFSReq req;
      icefsthrift::IcefsStatFSRes res;
      struct statvfs statData;

      req.__set_inode(inode);

      icefsThriftConn *thriftConn = thriftClientPool->GetIcefsThriftConn();
      thriftConn->thriftClient->DoIcefsStatFS(res, req);
      thriftClientPool->PutIcefsThriftConn(thriftConn);
      if (!res.status) {
        IcefsThriftFillStatvfsIn(&statData, res.statvfs);
        fuse_reply_statfs(fuseReq, &statData);
      } else {
        fuse_reply_err(fuseReq, res.status);
        ICEFS_PR_ERR_STATUS;
      }
      break;
    }

    default:
      fuse_reply_err(fuseReq, EIO);
      break;
  }
}
